test_numpy_object_array fails: TypeError: directly hashing ndarrays with dtype=object is unstable

**Describe the bug**
```

――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_numpy_object_array ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    def test_numpy_object_array():
        """
        _HASHABLE_EXTENSIONS = ub.util_hash._HASHABLE_EXTENSIONS
        """
        if np is None:
            pytest.skip('requires numpy')
        # An object array should have the same repr as a list of a tuple of data
        data = np.array([1, 2, 3], dtype=object)
>       objhash = ub.hash_data(data)

tests/test_hash.py:245: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
ubelt/util_hash.py:1107: in hash_data
    _update_hasher(hasher, data, types=types, extensions=extensions)
ubelt/util_hash.py:953: in _update_hasher
    prefix, hashable = _convert_to_hashable(data, types,
ubelt/util_hash.py:875: in _convert_to_hashable
    prefix, hashable = hash_func(data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

data = array([1, 2, 3], dtype=object)

    @self.register(np.ndarray)
    def _convert_numpy_array(data):
        """
        Example:
            >>> import ubelt as ub
            >>> if not ub.modname_to_modpath('numpy'):
            ...     raise pytest.skip()
            >>> import numpy as np
            >>> data_f32 = np.zeros((3, 3, 3), dtype=np.float64)
            >>> data_i64 = np.zeros((3, 3, 3), dtype=np.int64)
            >>> data_i32 = np.zeros((3, 3, 3), dtype=np.int32)
            >>> hash_f64 = _hashable_sequence(data_f32, types=True)
            >>> hash_i64 = _hashable_sequence(data_i64, types=True)
            >>> hash_i32 = _hashable_sequence(data_i64, types=True)
            >>> assert hash_i64 != hash_f64
            >>> assert hash_i64 != hash_i32
        """
        if data.dtype.kind == 'O':
            msg = 'directly hashing ndarrays with dtype=object is unstable'
>           raise TypeError(msg)
E           TypeError: directly hashing ndarrays with dtype=object is unstable

ubelt/util_hash.py:546: TypeError

```

Version: 1.3.0
Python-3.9
FreeBSD 13.2

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,7 +5,7 @@ We are currently working on porting this changelog to the specifications in
 This project (loosely) adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 
-## Version 1.3.0 - Unreleased
+## Version 1.3.0 - 2022-06-14
 
 ### Changed
 * Removed experimental warning from `Path.copy` and `Path.move`; they are now well enough tested.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,15 @@ We are currently working on porting this changelog to the specifications in
 This project (loosely) adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 
+## Version 1.3.1 - 2022-06-14
+
+
+### Fixed
+
+* Bug where the first call to `ub.hash_data` would error if it was given a
+  ndarray with object type.
+
+
 ## Version 1.3.0 - 2022-06-14
 
 ### Changed

--- a/tests/test_path.py
+++ b/tests/test_path.py
@@ -49,13 +49,16 @@ def test_augpath_dpath():
 
 
 def test_ensuredir_recreate():
+    import pytest
     base = ub.Path.appdir('ubelt/tests').ensuredir()
     folder = join(base, 'foo')
     member = join(folder, 'bar')
-    ub.ensuredir(folder, recreate=True)
+    with pytest.warns():
+        ub.ensuredir(folder, recreate=True)
     ub.ensuredir(member)
     assert exists(member)
-    ub.ensuredir(folder, recreate=True)
+    with pytest.warns():
+        ub.ensuredir(folder, recreate=True)
     assert not exists(member)
 
 

--- a/tests/test_path.py
+++ b/tests/test_path.py
@@ -53,11 +53,11 @@ def test_ensuredir_recreate():
     base = ub.Path.appdir('ubelt/tests').ensuredir()
     folder = join(base, 'foo')
     member = join(folder, 'bar')
-    with pytest.warns():
+    with pytest.warns(DeprecationWarning):
         ub.ensuredir(folder, recreate=True)
     ub.ensuredir(member)
     assert exists(member)
-    with pytest.warns():
+    with pytest.warns(DeprecationWarning):
         ub.ensuredir(folder, recreate=True)
     assert not exists(member)
 

--- a/tests/test_stream.py
+++ b/tests/test_stream.py
@@ -1,0 +1,13 @@
+
+
+def test_capture_stream_error():
+    import ubelt as ub
+    class DummyException(Exception):
+        ...
+    try:
+        with ub.CaptureStdout() as cap:
+            print('hello there')
+            raise DummyException
+    except DummyException:
+        ...
+    assert cap.text.startswith('hello there')

--- a/ubelt/__init__.py
+++ b/ubelt/__init__.py
@@ -24,7 +24,7 @@ Testing:
     xdoctest ubelt
 """
 
-__version__ = '1.3.0'
+__version__ = '1.3.1'
 
 # Deprecated functions
 from ubelt.util_platform import (

--- a/ubelt/util_cmd.py
+++ b/ubelt/util_cmd.py
@@ -156,14 +156,14 @@ def cmd(command, shell=False, detach=False, verbose=0, tee=None, cwd=None,
         command (str | List[str]):
             command string, tuple of executable and args, or shell command.
 
-        shell (bool, default=False):
-            if True, process is run in shell.
+        shell (bool):
+            if True, process is run in shell. Defaults to False.
 
-        detach (bool, default=False):
-            if True, process is detached and run in background.
+        detach (bool):
+            if True, process is detached and run in background. Defaults to False.
 
-        verbose (int, default=0):
-            verbosity mode. Can be 0, 1, 2, or 3.
+        verbose (int):
+            verbosity mode. Can be 0, 1, 2, or 3. Defaults to 0.
 
         tee (bool | None):
             if True, simultaneously writes to stdout while capturing output
@@ -177,23 +177,24 @@ def cmd(command, shell=False, detach=False, verbose=0, tee=None, cwd=None,
         env (Dict[str, str] | None):
             environment passed to Popen
 
-        tee_backend (str, default='auto'): backend for tee output.
+        tee_backend (str): backend for tee output.
             Valid choices are: "auto", "select" (POSIX only), and "thread".
+            Defaults to "auto".
 
-        check (bool, default=False):
+        check (bool):
             if True, check that the return code was zero before returning,
             otherwise raise a :class:`subprocess.CalledProcessError`.
-            Does nothing if detach is True.
+            Does nothing if detach is True.  Defaults to False.
 
-        system (bool, default=False):
+        system (bool):
             if True, most other considerations are dropped, and
             :func:`os.system` is used to execute the command in a platform
             dependant way. Other arguments such as env, tee, timeout, and shell
-            are all ignored.  (new in version 1.1.0)
+            are all ignored. Defaults to False. (New in version 1.1.0)
 
         timeout (float | None):
             If the process does not complete in ``timeout`` seconds, raise a
-            :class:`subprocess.TimeoutExpired`. (new in version 1.1.0).
+            :class:`subprocess.TimeoutExpired`. (New in version 1.1.0).
 
         capture (bool):
             if True, the stdout/stderr are captured and returned in the
@@ -216,15 +217,16 @@ def cmd(command, shell=False, detach=False, verbose=0, tee=None, cwd=None,
         what they would be on the command line.
 
     Related Work:
-        https://github.com/pycontribs/subprocess-tee
-        https://github.com/mortoray/shelljob
-        https://github.com/netinvent/command_runner
-        https://www.pyinvoke.org/prior-art.html
+        Similar to other libraries: [SubprocTee]_, [ShellJob]_, [CmdRunner]_, [PyInvoke]_.
 
     References:
         .. [SO_11495783] https://stackoverflow.com/questions/11495783/redirect-subprocess-stderr-to-stdout
         .. [SO_7729336] https://stackoverflow.com/questions/7729336/how-can-i-print-and-display-subprocess-stdout-and-stderr-output-without-distorti
         .. [SO_33560364] https://stackoverflow.com/questions/33560364/python-windows-parsing-command-lines-with-shlex
+        .. [SubprocTee] https://github.com/pycontribs/subprocess-tee
+        .. [ShellJob] https://github.com/mortoray/shelljob
+        .. [CmdRunner] https://github.com/netinvent/command_runner
+        .. [PyInvoke] https://www.pyinvoke.org/prior-art.html
 
     CommandLine:
         xdoctest -m ubelt.util_cmd cmd:6

--- a/ubelt/util_cmd.py
+++ b/ubelt/util_cmd.py
@@ -200,7 +200,7 @@ def cmd(command, shell=False, detach=False, verbose=0, tee=None, cwd=None,
             information dictionary. Ignored if detatch or system is True.
 
     Returns:
-        dict:
+        dict | CmdOutput:
             info - information about command status.
             if detach is False ``info`` contains captured standard out,
             standard error, and the return code

--- a/ubelt/util_cmd.pyi
+++ b/ubelt/util_cmd.pyi
@@ -8,6 +8,24 @@ POSIX: Incomplete
 WIN32: Incomplete
 
 
+class CmdOutput(dict):
+
+    @property
+    def stdout(self):
+        ...
+
+    @property
+    def stderr(self):
+        ...
+
+    @property
+    def returncode(self):
+        ...
+
+    def check_returncode(self) -> None:
+        ...
+
+
 def cmd(command: str | List[str],
         shell: bool = False,
         detach: bool = False,
@@ -18,5 +36,6 @@ def cmd(command: str | List[str],
         tee_backend: str = 'auto',
         check: bool = False,
         system: bool = False,
-        timeout: float | None = None) -> dict:
+        timeout: float | None = None,
+        capture: bool = True) -> dict | CmdOutput:
     ...

--- a/ubelt/util_format.py
+++ b/ubelt/util_format.py
@@ -24,7 +24,9 @@ def repr2(data, **kwargs):
         ...     'simple_list': [1, 2, 'red', 'blue'],
         ...     'odict': ub.odict([(2, '1'), (1, '2')]),
         ... }
-        >>> result = ub.repr2(dict_, nl=1, precision=2)
+        >>> import pytest
+        >>> with pytest.warns():
+        >>>     result = ub.repr2(dict_, nl=1, precision=2)
         >>> print(result)
         {
             'custom_types': [slice(0, 1, None), 0.33],

--- a/ubelt/util_format.py
+++ b/ubelt/util_format.py
@@ -25,7 +25,7 @@ def repr2(data, **kwargs):
         ...     'odict': ub.odict([(2, '1'), (1, '2')]),
         ... }
         >>> import pytest
-        >>> with pytest.warns():
+        >>> with pytest.warns(DeprecationWarning):
         >>>     result = ub.repr2(dict_, nl=1, precision=2)
         >>> print(result)
         {

--- a/ubelt/util_futures.py
+++ b/ubelt/util_futures.py
@@ -146,6 +146,15 @@ class SerialExecutor(object):
         return self
 
     def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         return False
 
     def submit(self, func, *args, **kw):
@@ -299,6 +308,15 @@ class Executor(object):
         return self
 
     def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         # Note: the following call will block
         return self.backend.__exit__(ex_type, ex_value, ex_traceback)
 
@@ -423,6 +441,15 @@ class JobPool(object):
         return self
 
     def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         return self.executor.__exit__(ex_type, ex_value, ex_traceback)
 
     def _clear_completed(self):
@@ -529,6 +556,10 @@ class JobPool(object):
         NOTE:
             The order of iteration may be changed in the future to be the
             submission order instead.
+
+        Yields:
+            concurrent.futures.Future:
+                The completed future object containing the results of a job.
 
         Example:
             >>> import ubelt as ub

--- a/ubelt/util_futures.pyi
+++ b/ubelt/util_futures.pyi
@@ -1,3 +1,5 @@
+from typing import Type
+from types import TracebackType
 import concurrent
 import concurrent.futures
 from typing import Callable
@@ -27,7 +29,9 @@ class SerialExecutor:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, ex_traceback):
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
     def submit(self, func, *args, **kw) -> concurrent.futures.Future:
@@ -50,7 +54,9 @@ class Executor:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, ex_traceback):
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
     def submit(self, func, *args, **kw) -> concurrent.futures.Future:
@@ -87,7 +93,9 @@ class JobPool:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, ex_traceback):
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
     def as_completed(
@@ -101,5 +109,5 @@ class JobPool:
     def join(self, **kwargs) -> List[Any]:
         ...
 
-    def __iter__(self):
+    def __iter__(self) -> concurrent.futures.Future:
         ...

--- a/ubelt/util_futures.pyi
+++ b/ubelt/util_futures.pyi
@@ -27,7 +27,7 @@ class SerialExecutor:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, tb) -> None:
+    def __exit__(self, ex_type, ex_value, ex_traceback):
         ...
 
     def submit(self, func, *args, **kw) -> concurrent.futures.Future:
@@ -50,7 +50,7 @@ class Executor:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, tb):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
         ...
 
     def submit(self, func, *args, **kw) -> concurrent.futures.Future:
@@ -87,7 +87,7 @@ class JobPool:
     def __enter__(self):
         ...
 
-    def __exit__(self, a, b, c) -> None:
+    def __exit__(self, ex_type, ex_value, ex_traceback):
         ...
 
     def as_completed(

--- a/ubelt/util_hash.py
+++ b/ubelt/util_hash.py
@@ -183,8 +183,8 @@ class _Hashers(object):
             try:
                 func()
             except ImportError:  # nocover
-                pass
-        self._lazy_queue = []
+                ...
+        self._lazy_queue.clear()
 
     def __contains__(self, key):
         if self._lazy_queue:  # nocover
@@ -338,6 +338,11 @@ class HashableExtensions(object):
         _hash_dispatch.__is_base__ = True
         self._hash_dispatch = singledispatch(_hash_dispatch)
 
+    def _evaluate_lazy_queue(self):
+        for func in self._lazy_queue:
+            func()
+        self._lazy_queue.clear()
+
     def register(self, hash_types):
         """
         Registers a function to generate a hash for data of the appropriate
@@ -489,10 +494,11 @@ class HashableExtensions(object):
             >>> print(f3(data))
         """
         # Evaluate the lazy queue if anything is in it
-        if self._lazy_queue:
-            for func in self._lazy_queue:
-                func()
-            self._lazy_queue = []
+        if self._lazy_queue:  # nocover
+            # Added nocover, because a bugfix prevents this from running and it
+            # is unclear how to build a test in for this.
+            self._evaluate_lazy_queue()
+
         query_hash_type = data.__class__
         # TODO: recognize some special dunder method instead
         # of strictly using this registry.
@@ -907,6 +913,11 @@ def _update_hasher(hasher, data, types=True, extensions=None):
     """
     if extensions is None:
         extensions = _HASHABLE_EXTENSIONS
+
+    # bugfix: ensure the lazy registration queue is evaluated before running
+    # the iterable checks.
+    if extensions._lazy_queue:
+        extensions._evaluate_lazy_queue()
 
     # Determine if the data should be hashed directly or iterated through
     if isinstance(data, (tuple, list, zip)):

--- a/ubelt/util_hash.py
+++ b/ubelt/util_hash.py
@@ -627,7 +627,7 @@ class HashableExtensions(object):
             >>> datas['udict_data2'] = ub.dict_isect(datas['udict_data1'], ['a', '4', 'c', 'b'])
             >>> datas['odict_data3'] = ub.dict_subset(datas['odict_data1'], ['c', 'b', 'a', '4'])
             >>> datas['udict_data3'] = ub.dict_isect(datas['udict_data1'], ['c', 'b', 'a', '4'])
-            >>> # print('datas = {}'.format(ub.repr2(datas, nl=-1)))
+            >>> # print('datas = {}'.format(ub.urepr(datas, nl=-1)))
             >>> for key, val in sorted(datas.items()):
             >>>     hashstr = ub.hash_data(val, base='abc', hasher='sha512', types=True)[0:8]
             >>>     print('{} = {}'.format(key, hashstr))

--- a/ubelt/util_import.py
+++ b/ubelt/util_import.py
@@ -86,7 +86,16 @@ class PythonPathContext(object):
             self.index = len(sys.path) + self.index + 1
         sys.path.insert(self.index, self.dpath)
 
-    def __exit__(self, type, value, trace):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         need_recover = False
         if len(sys.path) <= self.index:  # nocover
             msg_parts = [

--- a/ubelt/util_import.pyi
+++ b/ubelt/util_import.pyi
@@ -1,3 +1,5 @@
+from typing import Type
+from types import TracebackType
 from os import PathLike
 from types import ModuleType
 from typing import List
@@ -15,7 +17,9 @@ class PythonPathContext:
     def __enter__(self) -> None:
         ...
 
-    def __exit__(self, type, value, trace) -> None:
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
 

--- a/ubelt/util_path.py
+++ b/ubelt/util_path.py
@@ -385,7 +385,16 @@ class ChDir:
             os.chdir(self.context_dpath)
         return self
 
-    def __exit__(self, a, b, c):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         if self.context_dpath is not None:
             os.chdir(self.orig_dpath)
 
@@ -451,7 +460,16 @@ class TempDir:
     def __enter__(self):
         return self.start()
 
-    def __exit__(self, type_, value, trace):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         self.cleanup()
 
 

--- a/ubelt/util_path.pyi
+++ b/ubelt/util_path.pyi
@@ -1,5 +1,7 @@
 from os import PathLike
 from typing import Tuple
+from typing import Type
+from types import TracebackType
 from typing import List
 from typing import Callable
 from _typeshed import Incomplete
@@ -47,7 +49,9 @@ class ChDir:
     def __enter__(self):
         ...
 
-    def __exit__(self, a, b, c) -> None:
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
 
@@ -72,7 +76,9 @@ class TempDir:
     def __enter__(self):
         ...
 
-    def __exit__(self, type_, value, trace) -> None:
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...
 
 

--- a/ubelt/util_stream.py
+++ b/ubelt/util_stream.py
@@ -211,7 +211,16 @@ class CaptureStdout(CaptureStream):
         self.cap_stdout.close()
         self.cap_stdout = None
 
-    def __exit__(self, type_, value, trace):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         if self.enabled:
             try:
                 self.log_part()
@@ -219,5 +228,5 @@ class CaptureStdout(CaptureStream):
                 raise
             finally:
                 self.stop()
-        if trace is not None:
+        if ex_traceback is not None:
             return False  # return a falsey value on error

--- a/ubelt/util_stream.py
+++ b/ubelt/util_stream.py
@@ -224,8 +224,6 @@ class CaptureStdout(CaptureStream):
         if self.enabled:
             try:
                 self.log_part()
-            except Exception:  # nocover
-                raise
             finally:
                 self.stop()
         if ex_traceback is not None:

--- a/ubelt/util_stream.pyi
+++ b/ubelt/util_stream.pyi
@@ -1,4 +1,6 @@
 import io
+from typing import Type
+from types import TracebackType
 import io
 from _typeshed import Incomplete
 
@@ -61,5 +63,7 @@ class CaptureStdout(CaptureStream):
     def close(self) -> None:
         ...
 
-    def __exit__(self, type_, value, trace):
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...

--- a/ubelt/util_time.py
+++ b/ubelt/util_time.py
@@ -573,9 +573,18 @@ class Timer(object):
         self.tic()
         return self
 
-    def __exit__(self, ex_type, ex_value, trace):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         self.elapsed = self.toc()
-        if trace is not None:
+        if ex_traceback is not None:
             return False
 
 

--- a/ubelt/util_time.pyi
+++ b/ubelt/util_time.pyi
@@ -1,4 +1,6 @@
 import datetime
+from typing import Type
+from types import TracebackType
 from _typeshed import Incomplete
 
 
@@ -41,5 +43,7 @@ class Timer:
     def __enter__(self):
         ...
 
-    def __exit__(self, ex_type, ex_value, trace):
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...

--- a/ubelt/util_zip.py
+++ b/ubelt/util_zip.py
@@ -372,7 +372,16 @@ class zopen(NiceRepr):
     def __enter__(self):
         return self
 
-    def __exit__(self, *args):
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        """
+        Args:
+            ex_type (Type[BaseException] | None):
+            ex_value (BaseException | None):
+            ex_traceback (TracebackType | None):
+
+        Returns:
+            bool | None
+        """
         self.close()
 
     # TODO: Allow for navigating inside of the zipfile

--- a/ubelt/util_zip.pyi
+++ b/ubelt/util_zip.pyi
@@ -1,3 +1,5 @@
+from typing import Type
+from types import TracebackType
 from _typeshed import Incomplete
 from ubelt.util_mixins import NiceRepr
 
@@ -41,5 +43,7 @@ class zopen(NiceRepr):
     def __enter__(self):
         ...
 
-    def __exit__(self, *args) -> None:
+    def __exit__(self, ex_type: Type[BaseException] | None,
+                 ex_value: BaseException | None,
+                 ex_traceback: TracebackType | None) -> bool | None:
         ...

